File: source/CSharp/Interval.cs
using System;

namespace Exploring.Net.CSharp {
    public struct Interval {
        public Interval(double start, double finish) {
            this.start = start;
            this.finish = finish;
        }

        public double MidPoint => (start + finish) / 2.0;
        public double Size => Math.Abs(finish - start);

        public bool CrossesZero => Math.Sign(start) == -Math.Sign(finish) && Math.Sign(start) != 0;
        public bool Ascending => start < finish;

        public Interval LeftHalf => new Interval(start, MidPoint);
        public Interval RightHalf => new Interval(MidPoint, finish);
        public Interval Swap => new Interval(finish, start);
        public Interval Map(Func<double, double> function) => new Interval(function(start), function(finish));

        readonly double start;
        readonly double finish;
    }
}

File: source/FSharp/Interval.fs
namespace Exploring.Net.FSharp

module Interval = 

    type IntervalType = {Start: float; Finish: float}

    let create start finish = {Start = start; Finish = finish}

    let midpoint interval = (interval.Start + interval.Finish) / 2.0
    let size interval = abs(interval.Finish - interval.Start)

    let crossesZero interval = sign(interval.Start) = -sign(interval.Finish) && sign(interval.Start) <> 0
    let ascending interval = interval.Start < interval.Finish

    let leftHalf interval = create interval.Start (midpoint interval)
    let rightHalf interval = create (midpoint interval) interval.Finish
    let swap interval = create interval.Finish interval.Start
    let map aFunction interval  = create (aFunction interval.Start) (aFunction interval.Finish)
